src/dataflow-types/src/client.rs

```diff
@@ -961,28 +1018,204 @@ pub mod process_local {
+                            Ok(client) => {
+                                self.backoff = Duration::from_millis(10);
+                                self.state = GrpcComputedTcpConn::AwaitResponse(client);
+                            }
+                            Err(e) => {
+                                self.backoff = cmp::min(self.backoff * 2, Duration::from_secs(1));
+                                info!(
+                                    "Connection refused: {}. Backoff {}ms",
+                                    e,
+                                    self.backoff.as_millis()
+                                );
+                                self.state =
+                                    GrpcComputedTcpConn::Backoff(Instant::now() + self.backoff);
+                            }
+                        }
+                    }
+                    GrpcComputedTcpConn::AwaitResponse(client) => {
+                        // The channel size is a arbitrary, but it should be a
+                        // small bounded channel such that backpressure is applied early.
+                        let (tx, rx) = mpsc::channel(4);
+                        match client.handle(ReceiverStream::new(rx)).await {
+                            Ok(resp_rx_wrap) => {
+                                let resp_rx = resp_rx_wrap.into_inner();
+                                info!("GrpcComputedClient: connected");
+                                self.state = GrpcComputedTcpConn::Connected((tx, resp_rx));
+                            }
+                            Err(err) => {
+                                info!(
+                                    "GrpcComputedClient: Connection to {} refused: {}",
+                                    &self.addr, err
+                                );
+                                self.state = GrpcComputedTcpConn::Disconnected;
+                            }
+                        }
+                    }
+                    GrpcComputedTcpConn::Connected(_) => break,
+                    GrpcComputedTcpConn::Backoff(deadline) => {
+                        time::sleep_until(*deadline).await;
+                        self.state = GrpcComputedTcpConn::Disconnected;
+                    }
+                }
+            }
+        }
+    }
+
+    #[async_trait]
+    impl<T> GenericClient<ComputeCommand<T>, ComputeResponse<T>> for GrpcComputedClient
+    where
+        T: ProtoSerializableTimestamp + 'static + std::fmt::Debug + Send,
+    {
+        async fn send(&mut self, cmd: ComputeCommand<T>) -> Result<(), anyhow::Error> {
+            let sender = if let GrpcComputedTcpConn::Connected((sender, _)) = &self.state {
+                sender
+            } else {
+                return Err(anyhow::anyhow!("Sent into disconnected channel"));
+            };
+            sender
+                .send(ProtoSerializableTimestamp::compute_command_into_proto(cmd))
+                .await
```

Comment:
This `await` is problematic because it is not cancellation safe. This particular `send` needs to be cancellation safe because it's called within a `select`. It could happen that we silently drop messages here. The replicated client spawns a task for each replica. The task is driven by `select`: https://github.com/MaterializeInc/materialize/blob/78dd316bc58d1842d38356726d1d4a48d56b9a6e/src/dataflow-types/src/client/replicated.rs#L63

Reply:
good catch, I test if unbounded  solves the issue. Using unbounded doesn't seem quite right either, because we lose back pressure.

Reply:
Unfortunately doesnt resolve the issue. I changed all the queues to unbounded and it's still the same. But I'll chase down the select and see if I introduced other problematic cases...